Allow date filters to output ordinal days

Ana06 opened the pull request:
It is normal to write things like _posted on 12th February 2018" in the blog post. Currently to achieve this, a code similar to this one is needed:

```
{% assign d = page.date | date: "%-d" %}
{% case d %}
{% when '1' or '21' or '31' %}
{{ d }}st
{% when '2' or '22' %}
{{ d }}nd
{% when '3' or '23' %}
{{ d }}rd
{% else %}{{ d }}th
{% endcase %}
{{ page.date | date: "%B" }}
{{ page.date | date: "%Y" }}
```

I think the case is common enough to introduce it in Jekyll, by adding some parameters to the `date_to_string` and `date_to_long_string` methods.

It supports both UK and US styles. UK is the default.

I also changed @sample_date in one of the tests to cover more cases.

Ana06 commented:
I closed https://github.com/jekyll/jekyll/pull/6769 by accident

@ashmaroli I have made the changes we had discussed.

Rubocop complains about the class being to long, should I move date related method to a separate module, similarly to how it is done in `url_filters.rb`?

ashmaroli commented:
> should I move date related method to a separate module

I don't see the reason to *not move* this to a separate module. But I'll let the other maintainers make that decision..

DirtyF commented:
Are there no native methods in Ruby to handle ordinal dates? What about non-english dates format? e.g,  In french you write `Jeudi 1er Mars 2018`.

ashmaroli commented:
> Are there no native methods in Ruby to handle ordinal dates?

none that I'm aware of.. But interesting question..

Ana06 commented:
@DirtyF 

> Are there no native methods in Ruby to handle ordinal dates? What about non-english dates format? e.g, In french you write Jeudi 1er Mars 2018

In the [Time Ruby class](http://ruby-doc.org/core-2.5.0/Time.html) there is nothing. We could suggest them to add it, but it won't happen soon.

I think it is a bad idea to support dates in other languages in Jekyll, but what about extending this methods in [your Date plugin](https://github.com/envygeeks/jekyll-dates)? I would be up to help. 😉 

So my suggestion is add the ordinal dates in English as it is quite general and used case, and the rest in the plugin. What do you think? 😄

DirtyF commented:
@Ana06 OK, I guess it's make sense to implement ordinal dates like Rails:

https://github.com/rails/rails/blob/acbcef6094d61eb8c4820295620d170743a4bd71/activesupport/lib/active_support/inflector/methods.rb#L343-L369

DirtyF commented:
@Ana06 II mean I looked at your implementation and it seems to match the Rail's one 👍

mattr- commented on docs/_docs/templates.md:
```diff
@@ -105,6 +105,35 @@ you come up with your own tags via plugins.
+    <tr>
+      <td>
+        <p class="name"><strong>Date to String in ordinal US style</strong></p>
+        <p>Format a date to ordinal, UK, short format.</p>
```
Could you replace `UK` here with `US` since it looks like this is for the US format based on the context around it?

mattr- commented on lib/jekyll/filters.rb:
```diff
@@ -68,22 +68,33 @@ def slugify(input, mode = nil)
     end
 
     # Format a date in short format e.g. "27 Jan 2011".
+    # It allows to use ordinal format as well, in both UK
```
This comment would read better if it was written thusly:
```
Ordinal format is also supported, in both the UK and US formats.
```

Please keep the example dates in this comment when you make the language change. I left them out of the above sentence just to keep that shorter.

mattr- commented on lib/jekyll/filters.rb:
```diff
@@ -68,22 +68,33 @@ def slugify(input, mode = nil)
-      return date if date.to_s.empty?
-      time(date).strftime("%d %B %Y")
+    def date_to_long_string(date, type = nil, style = nil)
+      generic_date_to_string(date, "B", type, style)
```
Could you convert the magic string `B` into a constant? Same for `b` as well. Something like:
```ruby
LONG_MONTH_FORMAT = 'B'
SHORT_MONTH_FORMAT = 'b'
```

provides much more expressiveness and clarity in the code than the magic string.

Ana06 replied:
I changed it to @ashmaroli suggestion:

> instead of passing b and B as parameters, how about passing %b and %B ?

Do you think it is already clear?

I think with that is already clear that is the use of the strftime method and IMO there is no need to add two variables for this. :smile:

mattr- commented on lib/jekyll/filters.rb:
```diff
@@ -357,6 +368,34 @@ def sort_input(input, property, order)
+    # month_type: "b", "B"
+    # type: nil, "ordinal"
+    # style nil, "US"
+    def generic_date_to_string(date, month_type, type = nil, style = nil)
```
Both `type` and `style` should default to something other than `nil` here, IMHO.

Ana06 replied:
`nil` is equivalent to default. We can give it something more complicate, but the code will work anyway with something different. :thinking:

ashmaroli commented on docs/_docs/templates.md:
```diff
@@ -105,6 +105,35 @@ you come up with your own tags via plugins.
+      </td>
+    </tr>
+    <tr>
+    <tr>
```
double `<tr>` here..

ashmaroli commented on docs/_docs/templates.md:
```diff
@@ -105,6 +105,35 @@ you come up with your own tags via plugins.
+        <p class="name"><strong>Date to String in ordinal US style</strong></p>
+        <p>Format a date to ordinal, UK, short format.</p>
+      </td>
+      <td class="align-center">
```
Please **reorder** the table rows here such that information flows gracefully. For example:

Filter Desc. | Example
--- | --- 
**Date to String** | `date_to_string`
**Date to String in UK Ordinal** | `date_to_string: "ordinal"`
**Date to String in US Ordinal** | `date_to_string: "ordinal", "US"`
**Date to Long String** | `date_to_long_string`
**Date to Long String in UK Ordinal** | `date_to_long_string: "ordinal"`
**Date to Long String in US Ordinal** | `date_to_long_string: "ordinal", "US"`

ashmaroli commented on lib/jekyll/filters.rb:
```diff
@@ -68,22 +68,33 @@ def slugify(input, mode = nil)
-    def date_to_string(date)
-      time(date).strftime("%d %b %Y")
+    def date_to_string(date, type = nil, style = nil)
+      generic_date_to_string(date, "b", type, style)
```
- instead of `generic_date_to_string()` how about `stringify_date()` ?
- instead of passing `b` and `B` as parameters, how about passing `%b` and `%B` ?

DirtyF commented:
@Ana06 👋 we're still interested in shipping this for the next minor version, is there anything blocking you from addressing the feedback?

Ana06 commented:
@DirtyF 

> @Ana06 wave we're still interested in shipping this for the next minor version, is there anything blocking you from addressing the feedback?

I was waiting in case there was more feedback, I'll update this now :wink:

ashmaroli commented on lib/jekyll/filters.rb:
```diff
@@ -357,6 +368,34 @@ def sort_input(input, property, order)
+      if type == "ordinal"
+        day = time.day
+        ordinal_day = "#{day}#{ordinal(day)}"
+        return time.strftime("%#{month_type} #{ordinal_day}, %Y") if style == "US"
```
you forgot to adjust the string interpolation here and below.. :slightly_smiling_face: 
```diff
- time.strftime("%#{month_type}..
+ time.strftime("#{month_type}..
```

Ana06 replied:
yes, I was just correcting it :wink:

Ana06 commented:
@DirtyF I think it should be ready to merge now :wink:

Ana06 commented on lib/jekyll/filters/date_filters.rb:
```diff
@@ -66,21 +64,26 @@ def date_to_rfc822(date)
         time.strftime("%d #{month_type} %Y")
       end
 
+      private
```
:see_no_evil:

Ana06 commented on lib/jekyll/filters/date_filters.rb:
```diff
@@ -66,21 +64,26 @@ def date_to_rfc822(date)
-      # month_type: "b", "B"
-      # type: nil, "ordinal"
-      # style nil, "US"
+      # month_type: Notations that evaluate to 'Month' via `Time#strftime` ("%b", "%B")
```
:+1:

Ana06 commented:
@ashmaroli should I add your correction commit to mines, so that the commit history is clean? :thinking:

Ana06 commented on lib/jekyll/filters/date_filters.rb:
```diff
@@ -66,21 +64,26 @@ def date_to_rfc822(date)
-      # type: nil, "ordinal"
-      # style nil, "US"
+      # month_type: Notations that evaluate to 'Month' via `Time#strftime` ("%b", "%B")
+      #       type: nil (default) or "ordinal"
```
to many spaces....

ashmaroli commented:
Feel free to update the branch as you see fit. I only pushed the commit because I didn't want to bother you unnecessarily..

Ana06 commented:
@ashmaroli already added. I think I didn't miss anything :wink: I like having clean commit history :cupid:

ashmaroli commented:
> I think I didn't miss anything

no probs.. my changes seem to have been incorporated nicely.. :+1:

DirtyF commented:
📆 🤙 🔢 

Congrats @Ana06 for your first Jekyll contrib! 😍 

@jekyllbot: :shipit: +minor
